044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,275 @@
+future sources of quotas, like pull them from an external system.
+
+#### Throttle Factor
+
+We introduce the concept of throttle factor. This is a factor in the range (0.0, 1.0) that is applied to a quota to
+calculate the final byte rate limit. 
+
+This is a concept that was already implicit in the current calculations, but we want to name it so that we can use it in 
+metric names and provide an extension point in case we want to externalise the Throttle Factor.
+
+#### Throttle Factor Source
+
+We propose adding a Throttle Factor Source concept to the plugin to provide an extension point where we could plug in
+future sources of throttle factors, like pull factors from a single decision-making broker.
+
+#### Volume
+
+The storage quotas operate on observations about a **Volume** with these characteristics:
+1. `logDir`: the path of the logdir
+2. `brokerId`
+3. `totalBytes`
+4. `availableBytes`
+5. `usedBytes` (totalBytes - availableBytes)
+6. `availableRatio` (availableBytes/totalBytes)
+
+#### Volume Source
+
+The quota plugin operates on observations about Volumes, the **Volume Source** determines where
+we obtain those observations. Values are:
+1. `local`: we inspect this brokers local log dirs to discover the state of their Volume
+2. `cluster`: we ask kafka for the state of all log dirs for all currently active nodes
+
+the source will be configurable with a property like:
+
+1. `client.quota.callback.static.storage.volume.source=cluster`
+
+Local will be the default if the property isn't provided but is deprecated from the beginning.
+
+#### Cluster Volume Source
+
+With the introduction of [KIP-827](https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+log+dirs+total+and+usable+space+via+Kafka+API) in
+kafka 3.3 we can now obtain the total and usable (available) bytes per log dir as part of the DescribeLogDirsResponse.
+
+Note: if a single disk contains multiple log dirs, it will be described multiple times through the kafka APIs. This
+repetition is acceptable as our new limit types will be applied per-volume, so redundant volume descriptions don't
+impact the outcome.
+
+The Cluster Volume Source will use this API to discover volume information for the whole cluster.
+
+The performance cost of this API was [discussed](https://lists.apache.org/thread/11zyqqnyg1wgf4jdo6pvn7hn51g3vf8r) 
+upstream as part of the KIP, which should be low cost. The performance impact of polling is managed by making the poll
+interval configurable. The poll interval will be configurable using the existing `client.quota.callback.static.storage.check-interval` 
+property (default is 0 - disabled).
+
+##### Admin Client Configuration
+
+To obtain log dir descriptions though the admin api we need to construct an admin client. We intend to
+connect to the replication listener and will piggyback on the broker's existing SSL configuration to
```

Comment:
now that Strimzi has a dedicated control plane listener, would it be better to use that?  You could argue that what the plugin is doing is more like 'control' traffic than replication traffic.
 
 https://github.com/strimzi/proposals/blob/main/025-control-plain-listener.md

Reply:
So far the decision was to keep the controll plane listener only for Kafka. It is not used by other traffic of this nature either (Operators, Cruise Control etc.). Also, I think the general expectation is that with KRaft, brokers will not have the control plane listener.